### PR TITLE
Introduce the DB based index to the host tests

### PR DESCRIPTION
**Pull request by habdelra**
This PR integrates the SQLite based indexer into our host tests. Specifically when the environment variable `PG_INDEXER=true` is provided, then the host tests will use the DB based index instead of the in-memory index. 

The significant updates to the system include:
1. A new `Worker` abstraction. This is used to run indexing jobs. The worker will run the fastboot container that will be used perform the index. In the browser tests we use the `current-run.ts` module directly to generate the index, but when we port the `Worker` to node it will be the mechanism that consumes fastboot. The interesting side effect here is that the `Realm` and `SearchIndex` no longer have anything to do with fastboot. There is a `Worker` instance that is paired to each `Realm` instance, since the worker needs the `Realm`'s adapter in order to perform the indexing. To that end, there is a job category for each ream that search index instances publish indexing jobs to.
2. After fighting many bugs, it became clear that modules also need to participate in our index. Modules are currently already edges in our invalidation graph and not having them be represented in our index was becoming untenable. To that end, I did a bunch of renaming of the DB and columns to make it less card specific since index entries can be either modules or cards. I spend a good part of the day then diving thru the intricacies of node resolution for modules vs literal module URL's in the filesystem. The main summary is that in our `deps` column we used node resolution for our module names, but in our invalidation output (which feeds the SSE system) we use the full URL of modules as they appear in the file system. This distinction is important as we want consumers to be able to use node resolution for declaring their imports, but to have that bottom out on concrete files that drive live card/module events in our UI via SSE.

There is an interesting comment below around empty values vs missing values in our search index. I think we have been conflating this in our tests, but we need to get very concrete as there are SQLite limitations with the `json_tree()` function (literally the only real SQLite tool that we have in our arsenal to drive the our Query API). Ultimately I think we need to treat missing and empty as the same thing in our search index, and when serializing search docs, we need to materialize any missing fields into `null` values. much more detailed discussion is below.

At the time of this writing there are still about 7 failing tests. But i think the PR is far enough along that it could use some eyeballs. It's a big PR. not necessarily that many new lines of code, but lots of churned code. When we remove the feature flag, though, I think it will be about a break even amount of code, maybe even net negative.

### CODE REVIEW

File: packages/host/tests/integration/search-index-test.gts [hunk 1]
```diff
@@ -33,3907 +33,4076 @@ const testModuleRealm = 'http://localhost:4202/test/';
...
+        hooks.beforeEach(function (this: RenderingTestContext) {
+          (globalThis as any).__enablePgIndexer = () => isDbIndexingEnabled;
+          loader = (
+            this.owner.lookup('service:loader-service') as LoaderService
+          ).loader;
+        });
+
+        hooks.afterEach(function (this: RenderingTestContext) {
+          (globalThis as any).__enablePgIndexer = () => false;
+        });
```
Comment (habdelra) [inline packages/host/tests/integration/search-index-test.gts]: everything after this point is just indenting changes

File: packages/host/tests/integration/search-index-test.gts [hunk 2]
```diff
@@ -33,3972 +33,4064 @@ const testModuleRealm = 'http://localhost:4202/test/';
...
+              filter: {
+                type: { module: `${testModuleRealm}book`, name: 'Book' },
+              },
+            });
+            // note that sorting by nulls is problematic in that sqlite
+            // considers nulls the smallest possible value and postgres considers
+            // nulls the largest possible value. removing tests that make
+            // assertions around the positions of nulls as it cannot be run
+            // consistently between postgres, sqlite, and our in-memory index
+            assert.deepEqual(
```
Comment (habdelra) [inline packages/host/tests/integration/search-index-test.gts]: FYI making sorting assertions around `null` will be very difficult as there is not a consistent way nulls are treated during sorting. we should just rely on the documented DB behavior here for how nulls are sorted.

File: packages/host/tests/integration/search-index-test.gts [hunk 3]
```diff
@@ -33,3972 +33,4064 @@ const testModuleRealm = 'http://localhost:4202/test/';
...
+
+          // There is actually a sorting bug here in our original in-memory based
+          // index that was revealed when we started using the DB based index.
+          // probably not worth fixing as we are about to remove the in-memory based
+          // index shortly.
+          !isDbIndexingEnabled
+            ? skip(
+                "can sort on multiple paths in combination with 'any' filter",
+              )
+            : test(`can sort on multiple paths in combination with 'any' filter`, async function (assert) {
```
Comment (habdelra) [inline packages/host/tests/integration/search-index-test.gts]: FYI: caught an interesting sorting issue with multi-level sorting in the old code (the null sorting test assertions were hiding the bug). not sure its worth fixing as it is going away

File: packages/host/tests/integration/search-index-test.gts [hunk 4]
```diff
@@ -33,3972 +33,4064 @@ const testModuleRealm = 'http://localhost:4202/test/';
...
-            },
-            relationships: {
-              owner: {
-                links: {
-                  self: `${testRealmURL}Person/owner`,
+module('Integration | search-index', function (_hooks) {
+  // TODO run for both in-memory and db based indexing
+  for (let isDbIndexingEnabled of [false, true]) {
+    module(
+      isDbIndexingEnabled ? 'db index' : 'in-memory index',
```
Comment (habdelra) [inline packages/host/tests/integration/search-index-test.gts]: the primary change in this test is the indenting this `for` loop introduced

File: packages/runtime-common/indexer.ts [hunk 5]
```diff
@@ -147,7 +164,38 @@ export class Indexer {
       return undefined;
     }
 
-    return maybeResult;
+    if (maybeResult.error_doc) {
+      // TODO make a test for this!
+      return { type: 'error', error: maybeResult.error_doc };
+    }
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: don't forget!

File: packages/runtime-common/indexer.ts [hunk 6]
```diff
@@ -208,6 +256,7 @@ export class Indexer {
     let conditions: CardExpression[] = [
       ['i.realm_url = ', param(realmURL.href)],
       ['is_deleted = FALSE OR is_deleted IS NULL'],
+      ['error_doc IS NULL'], // TODO include a test for this!
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: don't forget!

File: packages/runtime-common/indexer.ts [hunk 7]
```diff
@@ -668,14 +714,30 @@ export class Indexer {
...
+      // cards grid sorts by the card's display name). current-run is injecting
+      // this into the searchDoc during index time.
+      field = {
+        card: (
+          await loader.import<{ default: typeof CardAPI.FieldDef }>(
+            'https://cardstack.com/base/string',
+          )
+        ).default,
+        fieldType: 'contains',
+      } as unknown as Field; // just pretend this is an actual field
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: FYI this was an interesting situation i ran into. hopefully there is not more data like this. if so we should probably make a `_meta` field in the search doc to hold this kind of stuff

File: packages/runtime-common/indexer.ts [hunk 8]
```diff
@@ -790,10 +853,15 @@ export class Batch {
     ]);
   }
 
+  // TODO I don't think we actually need this--we rely on the fact that deleted
+  // entries will be invalidated as part of incremental update, and if it is
+  // tru0ly gone, then we'll never visit the file as we try to visit the
+  // invalidations.
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts]: Don't forget to come back to this

File: packages/runtime-common/loader.ts [hunk 9]
```diff
@@ -435,10 +435,9 @@ export class Loader {
     urlOrRequest: string | URL | Request,
     init?: RequestInit,
   ): Promise<Response> {
+    let request = this.asRequest(urlOrRequest, init);
```
Comment (habdelra) [inline packages/runtime-common/loader.ts]: i needed to make this adjustment so that we could actually mutate the request in the loader url handlers (this lets us add a special header to the request). Otherwise any mutation you do to the request is just thrown away.

File: packages/runtime-common/search-index.ts [hunk 10]
```diff
@@ -254,6 +309,8 @@ export class SearchIndex {
     });
   }
 
+  // TODO I think we can break this out into a different module specifically a
+  // queue handler for incremental and fromScratch indexing
   private async setupRunner(start: () => Promise<void>) {
```
Comment (habdelra) [inline packages/runtime-common/search-index.ts:387]: don't forget to revisit this

File: packages/runtime-common/search-index.ts [hunk 11]
```diff
@@ -233,6 +287,7 @@ export class SearchIndex {
       if (!this.#incremental) {
         throw new Error(`Index runner has not been registered`);
       }
+      // TODO this should be published into the queue
       let current = await this.#incremental(
         this.#index,
         url,
```
Comment (habdelra) [inline packages/runtime-common/search-index.ts]: don't forget to do this!

File: packages/host/tests/integration/search-index-test.gts [hunk 12]
```diff
@@ -33,3972 +33,4063 @@ const testModuleRealm = 'http://localhost:4202/test/';
...
+              assert.ok(
+                false,
+                `search entry was an error: ${mango?.error.detail}`,
+              );
+            }
+          }
+        });
+
+        skip('can query the "production" index while performing indexing operations', async function (_assert) {});
+
```
Comment (habdelra) [inline packages/host/tests/integration/search-index-test.gts]: don't forget this!

File: packages/host/tests/integration/search-index-test.gts [hunk 13]
```diff
@@ -3122,6 +3216,7 @@ posts/ignore-me.json
                 title: 'post 1',
                 author: {
                   firstName: 'A',
+                  lastName: null,
```
Comment (habdelra) [inline packages/host/tests/integration/search-index-test.gts:3216]: We have some fundamental challenges searching for missing fields with SQLite and it's `json_tree()` function. Keep in mind there is a difference between "missing" and "empty" (a bunch of tests conflate these 2 ideas). An empty field is a field that has no value, In JSON this is represented as a `null`. A missing field is a field that is not actually present. In our cards these can either be a field that is `undefined`, or when serialized, because `undefined` is not a thing in JSON, this is a field that is omitted from the resulting document.

the way the `json_tree()` function works is that each item in the JSON object is a row in the result set (we can ellide rows that correspond to paths that we don't care about to make things more performant). the desired search result will be one of these rows (via a cross join)--we use the field's JSON path in the search doc to understand the context of the value that we are matching against. The issue is, if a field is simply missing, then there is no row in the result set to test our predicate against. The best we can do is search for an empty field, `null`--those we can find. But a missing field (bascially an `undefined`) is invisible to `json_tree()`. I think we might need a new Query filter `missing` to handle this. ChatGPT thinks there are probably things that we can do in SQLite to find these items, but the structure of the query is gonna be very different.

These tests all claimed they were trying to match `null`, but in reality these tests were actually testing against missing fields, not empty fields. I updated the test data to align with the test description. however, this does mean that there is a gap in our testing around matching against missing fields. This is further compounded by the fact that JSON has no support for `undefined`. So there is not a way to serialize a query for a missing field in JSON (which is the wire format for queries). Because of this I'm really leaning more on the idea of a new `missing` query filter. maybe something like:
```json
{
  "on": { "module": "http://test/cards/person", "name": "Person" },
  "missing": "address.city"
}
```
which searches for `Person` cards that have a missing `address.city` value. Note that this is not the same as having a `null` value for `address.city`. In that case we can just use an `eq` filter for `null`.

Probably we also need to balance this with how likely we think this feature would actually be used. In practice just a capability might rarely, if ever, be used. As we probably meant to match against empty and not missing...

Comment (habdelra) [reply packages/host/tests/integration/search-index-test.gts:3216]: Conversely, we might decide that missing and empty are essentially equivalent, and in that case we could assert any missing fields are actually `null` in our search docs--that would actually mitigate the issue that `json_tree()` runs into. This is probably more in line with how the tests are written right now. (but we can add tests that explicitly state "missing" and test that explicitly state "empty"

File: packages/runtime-common/realm.ts [hunk 14]
```diff
@@ -415,6 +430,7 @@ export class Realm {
     return this.#adapter.createJWT(claims, expiration, this.#realmSecretSeed);
   }
 
+  // TODO bypass the operationQueue when dbIndex is enabled
```
Comment (habdelra) [inline packages/runtime-common/realm.ts]: don't forget this!

File: packages/runtime-common/realm.ts [hunk 15]
```diff
@@ -500,6 +516,7 @@ export class Realm {
     return { isTracked: false, url };
   }
 
+  // TODO bypass the operationQueue when dbIndex is enabled
```
Comment (habdelra) [inline packages/runtime-common/realm.ts]: don't forget this!

File: packages/base/card-api.gts [hunk 16]
```diff
@@ -442,49 +449,54 @@ class ContainsMany<FieldT extends FieldDefConstructor>
...
         attributes: {
-          [this.name]: values.map((value) =>
-            callSerializeHook(this.card, value, doc, undefined, opts),
-          ),
+          [this.name]:
+            values === null
+              ? null
+              : values.map((value) =>
+                  callSerializeHook(this.card, value, doc, undefined, opts),
+                ),
```
Comment (habdelra) [inline packages/base/card-api.gts:457]: this is the actual change here--the rest is just indenting

File: packages/base/card-api.gts [hunk 17]
```diff
@@ -442,49 +449,54 @@ class ContainsMany<FieldT extends FieldDefConstructor>
...
-        }
-        if (resource.meta && Object.keys(resource.meta).length === 0) {
-          delete resource.meta;
-        }
-        return resource;
-      });
+      let serialized =
+        values === null
+          ? null
+          : values.map((value, index) => {
```
Comment (habdelra) [inline packages/base/card-api.gts:465]: this is the actual change here--the rest is just indenting

File: packages/host/tests/integration/search-index-test.gts [hunk 18]
```diff
@@ -3589,50 +3668,63 @@ posts/ignore-me.json
...
-            eq: {
-              'hosts.firstName': 'Arthur',
-              sponsors: null,
-            },
-          },
+    !isDbIndexingEnabled
+      ? skip(
+          `can filter on an array of primitive fields inside a containsMany using 'eq'`,
+        )
+      : test(`can filter on an array of primitive fields inside a containsMany using 'eq'`, async function (assert) {
```
Comment (habdelra) [inline packages/host/tests/integration/search-index-test.gts:3675]: yea! we can test this now! :tada:

File: packages/host/tests/integration/search-index-test.gts (hunk 13 above)
Comment (habdelra) [reply packages/host/tests/integration/search-index-test.gts:3216]: I created ticket CS-6762 to track this work

File: packages/runtime-common/indexer.ts [hunk 19]
```diff
@@ -264,9 +333,10 @@ export class Indexer {
           fieldQuery(s.by, s.on, false, 'sort'),
           ')',
           s.direction ?? 'asc',
+          'NULLS LAST',
```
Comment (habdelra) [inline packages/runtime-common/indexer.ts:346]: Don’t forget to test this! (Probably  this needs to be db index only test since the in-memory index doesn’t seem very consistent with this behavior for multiple for keys)

Comment (habdelra) [reply packages/runtime-common/indexer.ts:346]: done

File: packages/runtime-common/loader.ts [hunk 20]
```diff
@@ -435,11 +435,20 @@ export class Loader {
...
   ): Promise<Response> {
+    let mergedHeaders: HeadersInit =
+      urlOrRequest instanceof Request
+        ? headersFromRequest(urlOrRequest)
+        : init?.headers ?? {};
     try {
       for (let handler of this.urlHandlers) {
         let request = this.asRequest(urlOrRequest, init);
+        mergedHeaders = { ...mergedHeaders, ...headersFromRequest(request) };
+        setHeaders(request, mergedHeaders);
```
Comment (tintinthong) [inline packages/runtime-common/loader.ts:446]: ~~Is this inside the loop or outside?~~

Actually, is this needed?

Comment (habdelra) [reply packages/runtime-common/loader.ts:446]: It’s inside the loop. Each handler has the ability to contribute header, so we carry that forward. The challenge is that there is not a stable request object being mutated with each handler (which is how middleware normally works). I tried changing that to a single request object that passes thru all the handlers. But there are some situations where we read the body stream and in doing so you prevent the request from actually working (you “spent” the request). So I tried as light a touch as possible here gathering up all the headers after each request regardless if it was a different request instance that what we started with and merging requests into each request before the handler runs and gathering up any new headers after it runs.

File: packages/runtime-common/loader.ts [hunk 21]
```diff
@@ -435,11 +435,20 @@ export class Loader {
...
 
         let result = await handler(request);
+        // the handler is allowed to mutate the headers, so we merge any updated headers
+        mergedHeaders = { ...mergedHeaders, ...headersFromRequest(request) };
```
Comment (tintinthong) [inline packages/runtime-common/loader.ts:450]: If its just realm auth headers, then can we make it specific as what header is being set by the handler? I dono seems like something someone will mis-use

File: packages/runtime-common/loader.ts (hunk 20 above)
Comment (habdelra) [reply packages/runtime-common/loader.ts:446]: Absolutely positively yes it is needed!!!!!

in this PR we actually set a special header to tell the system the request need to use a WIP index instead of a production index. That header should be in combination with another headers url handlers provide. The expectation here is that this is middleware.

File: packages/runtime-common/loader.ts (hunk 21 above)
Comment (habdelra) [reply packages/runtime-common/loader.ts:450]: LOL, it’s not. it’s a completely different header that has nothing to do with auth and everything to do with indexing. Read comment above.

Comment (habdelra) [reply packages/runtime-common/loader.ts:450]: I think it’s a bad idea to put header specific logic at this level. Header specific logic needs to live in the url handler itself—otherwise the abstraction starts leaking out. Also you can’t predict what other custom headers we’ll need in the future. It should just be a normal part of our url handler middleware. Any mutation of the requests should be allowed to carry thru to other handlers.

File: packages/runtime-common/loader.ts (hunk 20 above)
Comment (tintinthong) [reply packages/runtime-common/loader.ts:446]: Oh I see it now `X-Boxel-Use-WIP-Index`

**Review by tintinthong — APPROVED** [review]
Not totally confident but I have run thru the code to look for gotchas and seems good . I also ran the test locally succesfully.

**Comment by habdelra** [comment]
> Not totally confident but I have run thru the code to look for gotchas and seems good . I also ran the test locally succesfully.

just to double check, when you ran the tests did you use `PG_INDEXER=true` env var when you started the host? that enables the DB index, otherwise you'll be testing the old in-memory index.

**Comment by tintinthong** [comment]
> > Not totally confident but I have run thru the code to look for gotchas and seems good . I also ran the test locally succesfully.
> 
> just to double check, when you ran the tests did you use `PG_INDEXER=true` env var when you started the host? that enables the DB index, otherwise you'll be testing the old in-memory index.

Yep

**Comment by habdelra** [comment]
I'm happy to continue with any updates based on feedback even tho this PR is closed. This PR is too painful to keep open for very long due to risk of conflicts since it is such a large PR.

File: packages/host/tests/integration/search-index-test.gts (hunk 13 above)
Comment (lukemelia) [reply packages/host/tests/integration/search-index-test.gts:3216]: I think treating missing and empty as equivalent seems OK
